File: core/extensions/sim/src/trader_sim.cpp
//
// Created by Keren Dong on 2019-07-11.
//

#include "trader_sim.h"

namespace kungfu
{
    namespace wingchun
    {
        namespace sim
        {
            TraderSim::TraderSim(bool low_latency, yijinjing::data::locator_ptr locator, const std::string &account_id, const std::string &json_config) :
                    Trader(low_latency, std::move(locator), SOURCE_SIM, account_id)
            {
                yijinjing::log::copy_log_settings(get_io_device()->get_home(), account_id);
                nlohmann::json config = nlohmann::json::parse(json_config);
                std::string match_mode =  config.value("match_mode", "custom");
                this->mode_ = get_mode(match_mode);
                if (this->mode_ == MatchMode::Custom)
                {
                    throw wingchun_error("user defined match mode not supported");
                }
            }

            bool TraderSim::insert_order(const yijinjing::event_ptr &event)
            {
                const msg::data::OrderInput &input = event->data<msg::data::OrderInput>();
                msg::data::Order order = {};
                order_from_input(input, order);
                order.insert_time = kungfu::yijinjing::time::now_in_nano();
                order.update_time = kungfu::yijinjing::time::now_in_nano();
                int min_vol = get_instrument_type(input.instrument_id, input.exchange_id) == InstrumentType::Stock ? 100 : 1;
                if (order.volume < min_vol)
                {
                    order.status = OrderStatus::Error;
                }
                else
                {
                    switch (mode_)
                    {
                        case MatchMode::Reject:
                        {
                            order.status = OrderStatus::Error;
                            break;
                        }
                        case MatchMode::Pend:
                        {
                            order.status = OrderStatus::Pending;
                            break;
                        }
                        case MatchMode::Cancel:
                        {
                            order.status = OrderStatus::Cancelled;
                            break;
                        }
                        case MatchMode::Fill:
                        {
                            order.status = OrderStatus::Filled;
                            order.volume_traded = order.volume;
                            order.volume_left = 0;
                            break;
                        }
                        case MatchMode::PartialFillAndCancel:
                        {
                            order.volume_traded = min_vol;
                            order.volume_left = order.volume - order.volume_traded;
                            order.status = order.volume_left == 0 ? OrderStatus::Filled : OrderStatus::PartialFilledNotActive;
                            break;
                        }
                        case MatchMode::PartialFill:
                        {
                            order.volume_traded = min_vol;
                            order.volume_left = order.volume - order.volume_traded;
                            order.status = order.volume_left == 0 ? OrderStatus::Filled : OrderStatus::PartialFilledActive;
                            break;
                        }
                        default:
                        {
                            throw wingchun_error("user defined match mode not supported");
                        }
                    }
                }

                if (order.volume_traded > 0)
                {
                    auto writer = get_writer(event->source());
                    msg::data::Trade &trade = writer->open_data<msg::data::Trade>(0, msg::type::Trade);
                    uint64_t trade_id = writer->current_frame_uid();
                    trade.trade_id = trade_id;
                    trade.order_id = input.order_id;
                    trade.trade_time = kungfu::yijinjing::time::now_in_nano();
                    strcpy(trade.instrument_id, input.instrument_id);
                    strcpy(trade.exchange_id, input.exchange_id);
                    strcpy(trade.account_id, input.account_id);
                    trade.side = input.side;
                    trade.offset = input.offset;
                    trade.price = input.limit_price; //TODO support market order
                    trade.volume = order.volume_traded;
                    writer->close_data();
                }

                if (not is_final_status(order.status))
                {
                    pending_orders_[order.order_id] = order;
                    oid2source_[order.order_id] = event->source();
                }
                get_writer(event->source())->write(0, msg::type::Order, order);
                return true;
            }

            bool TraderSim::cancel_order(const yijinjing::event_ptr &event)
            {
                auto& action = event->data<msg::data::OrderAction>();
                auto order_id = action.order_id;
                auto it = pending_orders_.find(order_id);
                if (it == pending_orders_.end())
                {
                    SPDLOG_ERROR("target order {} is finished", order_id);
                    return false;
                }
                else
                {
                    auto& order =  it->second;
                    order.status = order.volume_traded == 0 ? OrderStatus::Cancelled : OrderStatus::PartialFilledNotActive;
                    if (oid2source_.find(order_id) != oid2source_.end())
                    {
                        auto source = oid2source_[order_id];
                        get_writer(source)->write(0, msg::type::Order, order);
                        oid2source_.erase(order_id);
                    }
                    pending_orders_.erase(it);
                }
                return true;
            }
        }
    }
}

File: core/extensions/sim/trader_sim.py
from pykungfu import longfist as pylongfist
from pykungfu import yijinjing as yjj
from pykungfu import wingchun as pywingchun
import json
import kungfu.wingchun.utils as wc_utils
import kungfu.wingchun.msg as wc_msg
import os
import sys
import importlib
from dotted_dict import DottedDict
from collections import namedtuple


class MatchMode:
    Reject = "reject"
    Pend = "pend"
    Cancel = "cancel"
    PartialFillAndCancel = "partialfillandcancel"
    PartialFill = "partialfill"
    Fill = "fill"
    Custom = "custom"


OrderRecord = namedtuple("OrderRecord", ["source", "dest", "order"])


class TraderSim(pywingchun.Trader):
    def __init__(self, low_latency, locator, account_id, json_config):
        pywingchun.Trader.__init__(self, low_latency, locator, "sim", account_id)
        config = json.loads(json_config)
        self.match_mode = config.get("match_mode", MatchMode.Custom)

        self.ctx = DottedDict()
        self.ctx.orders = {}

        if self.match_mode == MatchMode.Custom:
            path = config.get("path")
            simulator_dir = os.path.dirname(path)
            name_no_ext = os.path.split(os.path.basename(path))
            sys.path.append(os.path.relpath(simulator_dir))
            impl = importlib.import_module(os.path.splitext(name_no_ext[1])[0])
            self.ctx.insert_order = getattr(impl, 'insert_order', lambda ctx, event: False)
            self.ctx.cancel_order = getattr(impl, "cancel_order", lambda ctx, event: False)
            self.ctx.req_account = getattr(impl, "req_account", lambda ctx: False)
            self.ctx.req_position = getattr(impl, "req_position", lambda ctx: False)

    def on_start(self):
        pywingchun.Trader.on_start(self)

    def insert_order(self, event):
        if self.match_mode == MatchMode.Custom:
            return self.ctx.insert_order(self.ctx, event)
        else:
            order_input = event.data
            order = pywingchun.utils.order_from_input(order_input)
            min_vol = 100 if wc_utils.get_instrument_type(order_input.instrument_id,
                                                          order_input.exchange_id) == pylongfist.enums.InstrumentType.Stock else 1
            if order_input.volume < min_vol:
                order.status = pylongfist.enums.OrderStatus.Error
            elif self.match_mode == MatchMode.Reject:
                order.status = pylongfist.enums.OrderStatus.Error
            elif self.match_mode == MatchMode.Pend:
                order.status = pylongfist.enums.OrderStatus.Pending
            elif self.match_mode == MatchMode.Cancel:
                order.status = pylongfist.enums.OrderStatus.Cancelled
            elif self.match_mode == MatchMode.PartialFillAndCancel:
                order.volume_traded = min_vol
                order.status = pylongfist.enums.OrderStatus.Filled if order.volume_traded == order.volume \
                    else pylongfist.enums.OrderStatus.PartialFilledNotActive
            elif self.match_mode == MatchMode.PartialFill:
                order.volume_traded = min_vol
                order.status = pylongfist.enums.OrderStatus.Filled if order.volume_traded == order.volume \
                    else pylongfist.enums.OrderStatus.PartialFilledActive
            elif self.match_mode == MatchMode.Fill:
                order.volume_traded = order_input.volume
            else:
                raise Exception("invalid match mode {}".format(self.match_mode))
            order.volume_left = order.volume - order.volume_traded
            self.get_writer(event.source).write_data(0, wc_msg.Order, order)
            if order.volume_traded > 0:
                trade = pywingchun.Trade()
                trade.account_id = self.io_device.home.name
                trade.order_id = order.order_id
                trade.volume = order.volume_traded
                trade.price = order.limit_price
                trade.instrument_id = order.instrument_id
                trade.exchange_id = order.exchange_id
                trade.trade_id = self.get_writer(event.source).current_frame_uid()
                self.get_writer(event.source).write_data(0, wc_msg.Trade, trade)
            if order.active:
                self.ctx.orders[order.order_id] = OrderRecord(source=event.source, dest=event.dest, order=order)
            return True

    def cancel_order(self, event):
        if self.match_mode == MatchMode.Custom:
            return self.ctx.cancel_order(self.ctx, event)
        else:
            order_action = event.data
            if order_action.order_id in self.ctx.orders:
                record = self.ctx.orders.pop(order_action.order_id)
                order = record.order
                order.status = pylongfist.enums.OrderStatus.Cancelled if order.volume_traded == 0 \
                    else pylongfist.enums.OrderStatus.PartialFilledNotActive
                self.get_writer(event.source).write_data(0, wc_msg.Order, order)
            return True

    def re_account(self):
        if self.match_mode == MatchMode.Custom:
            return self.ctx.req_account(self.ctx)
        return False

    def req_position(self):
        if self.match_mode == MatchMode.Custom:
            return self.ctx.req_position(self.ctx)
        return False
